cs/src/core/Serializer.cs

```diff
@@ -76,10 +77,23 @@ public class Serializer<W>
         public Serializer(Type type, bool inlineNested)
         {
             var parser = new ObjectParser(type);
-            serialize = SerializerGeneratorFactory<object, W>.Create(
-                    (o, w, i) => serialize[i](o, w), type, inlineNested)
-                .Generate(parser)
-                .Select(lambda => lambda.Compile()).ToArray();
+
+            var firstPassAttribute = typeof(W).GetAttribute<FirstPassWriterAttribute>();
+            if (firstPassAttribute != null)
+            {
+                if (!typeof(ITwoPassProtocolWriter).IsAssignableFrom(typeof(W)))
+                {
+                    throw new ArgumentException("Writers with FirstPassWriterAttribute must implement ITwoPassProtocolWriter");
+                }
+
+                Type firstPassHelperType = typeof(TwoPassSerializerHelper<>).MakeGenericType(typeof(W), firstPassAttribute.Type);
+
+                helper = (SerializerHelper)Activator.CreateInstance(firstPassHelperType, parser, type, inlineNested);
+            }
+            else
+            {
+                helper = new SerializerHelper(parser, type, inlineNested);
+            }
```

Comment:
1. The logic from line 80 to line 96 just wants to be extracted into its own method; and it suspiciously looks very similar with the one in Transcoder.  
2. Also, it should be called only once per type `W` - please push it to the static constructor of Serializer<W>, especially since it calls Activator.CreateInstance